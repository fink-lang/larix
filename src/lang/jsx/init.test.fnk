{describe, it, expect, to_match_snapshot} = import '@fink/jest/test.fnk'

{parse_expr, parse} = import '../../parser.fnk'


describe 'JSX', fn:

  it 'parses self closing elem: <Foobar />', fn:
    expect
      parse_expr '<Foobar />'
      to_match_snapshot


  it 'parses self closing elem with short attr: <Foobar spam />', fn:
    expect
      parse_expr '<Foobar spam />'
      to_match_snapshot


  it 'parses self closing elem with str attr: <Foobar spam="ni" />', fn:
    expect
      parse_expr '<Foobar spam="ni" />'
      to_match_snapshot


  it 'parses self closing elem with str attr: <Foobar spam=\'ni\' />', fn:
    expect
      parse_expr '<Foobar spam="ni" />'
      to_match_snapshot


  it 'parses self closing elem with expr attr: <Foobar spam={ni} />', fn:
    expect
      parse_expr '<Foobar spam={ni} />'
      to_match_snapshot


  it 'parses empty elem: <Foobar></Foobar>', fn:
    expect
      parse_expr '<Foobar></Foobar>'
      to_match_snapshot


  it 'parses elem with children: <Foobar><Spam /></Foobar>', fn:
    expect
      parse_expr '
        <Foobar ni>
          <Spam />
        </Foobar>
      '
      to_match_snapshot


  it 'parses elem with expr in body: <Foobar> {...} </Foobar>', fn:
    expect
      parse_expr '
        <Foobar>
          {1 + 2}
        </Foobar>
      '
      to_match_snapshot


  it 'parses fragment', fn:
    expect
      parse_expr '
        <>
          {1 + 2}
        </>
      '
      to_match_snapshot

    expect
      parse_expr '
        <></>
      '
      to_match_snapshot



describe 'JSX parse failures', fn:
  it 'errors with invalid attr value', fn:
    expect
      parse '<Foobar spam=123 />', 'test.fnk'
      to_match_snapshot
