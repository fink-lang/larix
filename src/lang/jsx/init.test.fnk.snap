// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JSX parse failures errors with invalid attr value 1`] = `
module (1:0-1:19)
  jsx:elem (1:0-1:19) Foobar
    jsx:attr (1:8-1:13) spam
            ::error::
    jsx:attr (1:13-1:16) 123

error:
test.fnk:1:13
1| <Foobar spam=123 />
                ^

Expected one of [ '{', '"', "'" ] but found '123'.



`;

exports[`JSX parses elem with children: <Foobar><Spam /></Foobar> 1`] = `
jsx:elem (1:0-3:9) Foobar
  jsx:attr (1:8-1:10) ni
  :
  jsx:text (1:11-2:2)
    "\\n  "
  jsx:elem (2:2-2:10) Spam
  jsx:text (2:10-3:0)
    "\\n"
`;

exports[`JSX parses elem with expr in body: <Foobar> {...} </Foobar> 1`] = `
jsx:elem (1:0-3:9) Foobar
  :
  jsx:text (1:8-2:2)
    "\\n  "
  jsx:expr (2:2-2:9)
    block (2:2-2:9)
      arithm + (2:3-2:8)
        number (2:3-2:4) 1
        number (2:7-2:8) 2
  jsx:text (2:9-3:0)
    "\\n"
`;

exports[`JSX parses empty elem: <Foobar></Foobar> 1`] = `jsx:elem (1:0-1:17) Foobar`;

exports[`JSX parses fragment 1`] = `
jsx:frag (1:0-3:3) 
  :
  jsx:text (1:2-2:2)
    "\\n  "
  jsx:expr (2:2-2:9)
    block (2:2-2:9)
      arithm + (2:3-2:8)
        number (2:3-2:4) 1
        number (2:7-2:8) 2
  jsx:text (2:9-3:0)
    "\\n"
`;

exports[`JSX parses fragment 2`] = `jsx:frag (1:0-1:5) `;

exports[`JSX parses self closing elem with expr attr: <Foobar spam={ni} /> 1`] = `
jsx:elem (1:0-1:20) Foobar
  jsx:attr (1:8-1:17) spam
        jsx:expr (1:13-1:17)
      block (1:13-1:17)
        ident (1:14-1:16) ni
`;

exports[`JSX parses self closing elem with short attr: <Foobar spam /> 1`] = `
jsx:elem (1:0-1:15) Foobar
  jsx:attr (1:8-1:12) spam
`;

exports[`JSX parses self closing elem with str attr: <Foobar spam="ni" /> 1`] = `
jsx:elem (1:0-1:20) Foobar
  jsx:attr (1:8-1:17) spam
        jsx:string (1:14-1:16)
      'ni'
`;

exports[`JSX parses self closing elem with str attr: <Foobar spam='ni' /> 1`] = `
jsx:elem (1:0-1:20) Foobar
  jsx:attr (1:8-1:17) spam
        jsx:string (1:14-1:16)
      'ni'
`;

exports[`JSX parses self closing elem: <Foobar /> 1`] = `jsx:elem (1:0-1:10) Foobar`;
